_posts/on-viewmodel-composition-and-ui-composition.md

```diff
@@ -0,0 +1,152 @@
+   <products-list>
+      <item>
+         <product-name-and-description />
+         <product-price />
+      </item>
+      <item>
+         <product-name-and-description />
+         <product-price />
+      </item>
+   </products-list>
+<section>
+```
+
+You can probably quickly spot the issue. If each component is responsible for retrieving data for itself, the above snippet serves a "retrieve data N+1" scenario on a silver plate. For the sake of the discussion, let's imagine that each component uses HTTP to get data. In the above snippet, the user interface lists two products, each one composed of two components. When rendered, the UI will issue 4 HTTP requests, one for each component. Each component will hit its backend storage to retrieve data one by one. It's not that different from selecting data from a database without using joins or, worse, by issuing a select statement for each row that we need to load; this is when ViewModel Composition comes into play.
+
+## ViewModel Composition to the rescue
+
+ViewModel Composition is responsible for retrieving and composing data from different sources later presented to a consumer as a whole. The consumer at this point can be anything; the ViewModel Composition infrastructure doesn't care who the consumer is. Suppose the consumer is a user interface that uses UI Composition. In that case, ViewModel Composition lifts the above-described problems and is responsible for making sure that each request hits the backend infrastructure the fewest number of times. In the case of a distributed system, it means that, regardless of the number of items displayed on the UI, the number of backend requests should be not more than the number of services involved in the composition process.
+
+Using the product list sample presented above and expanding a bit to make it more real-world, let's imagine designing a home page where the need is to highlight the most popular products. Marketing is the service responsible for tracking products' popularity, and the only thing that stores are the product ID and a rank value. Catalog stores the product name and description, indexed by product ID. Finally, the Sales service stores the product price. All the data consumers require must be composed using not more than three requests, one for each service.
+
+ViewModel Composition techniques allow services to participate in the composition process to optimize the number of requests going to the backend. ViewModel Composition also comes with the advantage of masquerading the backend topology from the consumers. In most cases, ViewModel Composition uses a reverse proxy-like approach, being it a proper reverse proxy or a logical one, making so that consumers deal with a single source, the composition infrastructure, that returns a single cohesive and coherent ViewModel. From the consumer perspective, the composition infrastructure is entirely transparent; consumers won't realize that their data comes from multiple sources.
```

Comment:
```suggestion
ViewModel Composition techniques allow services to participate in the composition process to optimize the number of requests going to the backend. ViewModel Composition also comes with the advantage of masking the backend topology from the consumers. In most cases, ViewModel Composition uses a reverse proxy-like approach, being it a proper reverse proxy or a logical one, so that consumers deal with a single source, the composition infrastructure, that returns a single cohesive and coherent ViewModel. From the consumer perspective, the composition infrastructure is entirely transparent; consumers won't realize that their data comes from multiple sources.
```